reviewer/reviewer.go

```diff
@@ -0,0 +1,163 @@
+	for _, n := range u {
+		m[n] = struct{}{}
+	}
+	return m
+}
+
+func selectTeamMembers(prctx pull.Context, allTeams []string, r *rand.Rand) ([]string, error) {
+	randomTeam := allTeams[r.Intn(len(allTeams))]
+	teamInfo := strings.Split(randomTeam, "/")
+	teamMembers, err := prctx.ListTeamMembers(teamInfo[0], teamInfo[1])
+	if err != nil {
+		return nil, errors.Wrapf(err, "Unable to get member listing for team %s", randomTeam)
+	}
+	return teamMembers, nil
+}
+
+func FindRandomRequesters(ctx context.Context, prctx pull.Context, result common.Result, r *rand.Rand) ([]string, error) {
+	logger := zerolog.Ctx(ctx)
+	pendingLeafNodes := findLeafChildren(result)
+	var requestedUsers []string
+
+	logger.Debug().Msgf("Collecting reviewers for %d pending leaf nodes", len(pendingLeafNodes))
+
+	for _, child := range pendingLeafNodes {
+		allUsers := make(map[string]struct{})
+		allUsers = shoveIntoMap(allUsers, child.Rule.RequestedUsers)
+
+		if len(child.Rule.RequestedTeams) > 0 {
+			teamMembers, err := selectTeamMembers(prctx, child.Rule.RequestedTeams, r)
+			if err != nil {
+				logger.Warn().Err(err).Msgf("Unable to get member listing for teams, skipping team member selection")
+			}
+			allUsers = shoveIntoMap(allUsers, teamMembers)
+		}
+
+		if len(child.Rule.RequestedOrganizations) > 0 {
+			randomOrg := child.Rule.RequestedOrganizations[r.Intn(len(child.Rule.RequestedOrganizations))]
+			orgMembers, err := prctx.ListOrganizationMembers(randomOrg)
+			if err != nil {
+				logger.Warn().Err(err).Msgf("Unable to get member listing for org %s, skipping org member selection", randomOrg)
+			}
+			allUsers = shoveIntoMap(allUsers, orgMembers)
+		}
+
+		allCollaboratorPermissions, err := prctx.ListRepositoryCollaborators()
+		if err != nil {
+			return nil, errors.Wrap(err, "Unable to list repository collaborators")
+		}
+
+		if child.Rule.RequestedAdmins {
+			var repoAdmins []string
+			for _, c := range allCollaboratorPermissions {
+				if allCollaboratorPermissions[c] == "admin" {
+					repoAdmins = append(repoAdmins, c)
+				}
+			}
+			allUsers = shoveIntoMap(allUsers, repoAdmins)
+		}
+
+		if child.Rule.RequestedWriteCollaborators {
+			var repoCollaborators []string
+			for _, c := range allCollaboratorPermissions {
+				if allCollaboratorPermissions[c] == "write" {
+					repoCollaborators = append(repoCollaborators, c)
+				}
+			}
+			allUsers = shoveIntoMap(allUsers, repoCollaborators)
+		}
+
+		// Remove author before randomly selecting, since github will fail to assign _anyone_
+		if _, ok := allUsers[prctx.Author()]; ok {
+			delete(allUsers, prctx.Author())
+		}
+
+		// Remove any users who aren't collaborators, since github will fail to assign _anyone_
+		for k := range allUsers {
+			if _, ok := allCollaboratorPermissions[k]; !ok {
+				delete(allUsers, k)
+			}
+		}
+
+		var allUserList []string
+		for k := range allUsers {
+			allUserList = append(allUserList, k)
+		}
```

Comment:
Looking at this now, I think it would be simpler and more efficient to build the list in a single loop, where you append who are not the author and who are in the collaborators list.